### PR TITLE
Configuration option for nameserver and preflight check for loopback nameserver

### PR DESCRIPTION
**Pull request by areed**

### CODE REVIEW

**Review by paha — APPROVED** [review]

File: scripts/common/preflights.sh [hunk 1]
```diff
@@ -233,3 +234,15 @@ function kotsadm_prerelease() {
         fi
     fi
 }
+
+function host_nameservers_reachable() {
+    if [ "$AIRGAP" = 1 ]; then
+        return 0
+    fi
+    if ! discover_non_loopback_nameservers; then
+        printf "\n${YELLOW}No nameservers configured on the host are accessible by pods. All DNS queries outside the cluster will fail. Continue?${NC} "
```
Comment (marccampbell) [inline scripts/common/preflights.sh:243]: @areed can you help me understand how i'd want to answer this?  if i'm installing an application (the primary kURL use case today), in an airgapped environment, what does this mean?  If I see this, should I continue or abort?

File: scripts/common/preflights.sh [hunk 2]
```diff
@@ -233,3 +234,15 @@ function kotsadm_prerelease() {
...
+        if ! confirmN; then
+            bail "\nAt least one nameserver must be accessible on a non-loopback address"
```
Comment (marccampbell) [inline scripts/common/preflights.sh:245]: "accessible".  Does this mean accessibility or configured?

**Comment by marccampbell** [comment]
Is it possible to install a local DNS server on a server, configure resolv.conf to point to that and have the local DNS server be a cache through to upstream? Would this incorrectly warn that it's not set up properly?

Would a better solution be to test for the condition that we are looking for? This currently tests for the configuration and makes assumptions about what that means, when it might be easier or more reliable to test that DNS resolution works, using one of the known-open, documented Replicated endpoints?

File: scripts/common/preflights.sh (hunk 1 above)
Comment (areed) [reply scripts/common/preflights.sh:243]: In an airgapped environment this preflight is skipped. In an online environment you should abort and fix your nameserver.

File: scripts/common/preflights.sh (hunk 2 above)
Comment (areed) [reply scripts/common/preflights.sh:245]: The address in resolv.conf must be non-loopback. It's about what address the client uses not what address the nameserver binds.

**Comment by areed** [comment]
> Is it possible to install a local DNS server on a server, configure resolv.conf to point to that and have the local DNS server be a cache through to upstream? Would this incorrectly warn that it's not set up properly?

Yes, you can run dnsmasq but if that's your only nameserver in resolv.conf it has to be on the host IP, not loopback.

**Comment by areed** [comment]
> Would a better solution be to test for the condition that we are looking for? This currently tests for the configuration and makes assumptions about what that means, when it might be easier or more reliable to test that DNS resolution works, using one of the known-open, documented Replicated endpoints?

We could do that but then it wouldn't be a preflight, we'd have to wait for kubernetes to be installed

**Review by laverya — APPROVED** [review]

File: testgrid/tgrun/pkg/instances/k8s119_nameserver.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+		types.Instance{
+			InstallerSpec: types.InstallerSpec{
+				Kubernetes: kurlv1beta1.Kubernetes{
+					Version: "1.19.3",
+				},
+				Weave: &kurlv1beta1.Weave{
+					Version: "2.6.4",
+				},
+				Rook: &kurlv1beta1.Rook{
+					Version: "1.0.4",
```
Comment (laverya) [inline testgrid/tgrun/pkg/instances/k8s119_nameserver.go:19]: do we need rook here?
just want to keep tests minimal (and thus theoretically faster), and rook 1.0.4 is definitely tested elsewhere

**Review by laverya — APPROVED** [review]

File: testgrid/tgrun/pkg/instances/k8s119_nameserver.go (hunk 3 above)
Comment (areed) [reply testgrid/tgrun/pkg/instances/k8s119_nameserver.go:19]: No, I'll remove

**Pull request merged**